astoria/managers/astprocd.py

```diff
@@ -92,6 +108,51 @@ async def handle_disk_removal(self, uuid: DiskUUID, disk_info: DiskInfo) -> None
             else:
                 LOGGER.warning("Disk removed, but no code lifecycle available")
 
+    async def handle_kill_request(
+        self,
+        request: UsercodeKillManagerRequest,
+    ) -> RequestResponse:
+        """Handle a request to kill running usercode."""
+        if self._lifecycle is None:
+            return RequestResponse(
+                uuid=request.uuid,
+                success=False,
+                reason="No active usercode lifecycle",
+            )
+        else:
+            LOGGER.info("Kill request received.")
+            await self._lifecycle.kill_process()
+            return RequestResponse(
+                uuid=request.uuid,
+                success=True,
+            )
+
+    async def handle_restart_request(
+        self,
+        request: UsercodeRestartManagerRequest,
+    ) -> RequestResponse:
+        """Handle a request to restart usercode."""
+        LOGGER.info("Restart request received.")
+        if self._lifecycle is None:
+            return RequestResponse(
+                uuid=request.uuid,
+                success=False,
+                reason="No active usercode lifecycle",
+            )
+        else:
+            if self._lifecycle.status is CodeStatus.RUNNING:
+                return RequestResponse(
+                    uuid=request.uuid,
+                    success=False,
+                    reason="Code is already running.",
+                )
+            else:
+                asyncio.ensure_future(self._lifecycle.run_process())
```

Comment:
this feels not great

if someone e.g. hits the "start" button in the web ui twice quickly, it seems like this line could end up running twice (which would then mean that you have two `run_process` coroutines running two processes and sharing the same global state, which would be Bad)

imo the best way to fix this is probably in `run_process`, by replacing the `if self._process is None:` with something that atomically takes a lock or fails if the lock is already taken (I don't know what the idiomatic way to do that with asyncio is, sorry, probably `async with` an `asyncio.Lock`?), so that it is guaranteed that only one `run_process` will ever attempt to execute code at once

there is always the possibility I am misremembering how python's async model works? the scenario I'm imagining is:

- restart request received, `handle_restart_request` runs, schedules a `run_process`
- another restart request received, `handle_restart_request` runs, schedules a `run_process`
- `run_process` executes, `self._process` is None, awaits `create_subprocess_exec`
- the other `run_process` executes, `self._process` is still None, awaits a second `create_subprocess_exec`
- oh dear

Reply:
552a5b1